softwareComponents/shapeReconfig/include/shapeReconfig/algorithms.hpp

```diff
@@ -0,0 +1,934 @@
+ * BFShape - BFS with complete shape equality
+ * BFSEigen - BFS differentiating RoFIBots only by the eigenvalues of 
+ * their PCA decompositions; considers reflected shapes to be equal
+ * ShapeStar - A* using complete shape symmetry and a heuristic based on
+ * recognition of the shapes of individual modules of the RoFIBot
+ */
+enum class Algorithm { BFStrict, BFShape, BFSEigen, ShapeStar };
+/**
+ * Node type describes what each node in the state space is defined by:
+ * World - the corresponding RoFIWorld
+ * Shape - the shape of the corresponding RoFIWorld in cannonical form ("largest" Cloud)
+ * (unused, probably defines the same state space as EigenCloud, but is slower in experiments)
+ * Eigen - the eigenvalues of the PCA decomposition of the RoFIWorld
+ * EigenCloud - eigenvalues define the approximate shape, a (noncannonical) cloud
+ * is used for comparing the definitive shape (quickly decides non-equality of two nodes)
+ */
+enum class NodeType { World, Shape, Eigen, EigenCloud };
+
+struct Node;
+
+using NodeId = size_t;
+
+class Reporter;
+
+template < NodeType _NodeType >
+std::vector< RofiWorld > bfs( 
+    const RofiWorld& start, const RofiWorld& target,
+    float step, Reporter& rep, size_t maxDepth = 0 );
+
+template < NodeType _NodeType >
+std::vector< RofiWorld > shapeStar( 
+    const RofiWorld& start, const RofiWorld& target,
+    float step, Reporter& rep );
+
+template < NodeType _NodeType >
+std::vector< Node > bfsTraverse( const RofiWorld& start, float step, size_t maxDepth = 0 );
+
+struct Node
+{
+    NodeId nid;
+    RofiWorld world;
+    Cloud shape;
+    std::array< int, 4 > eigenVals;
+    size_t distFromStart;
+    size_t predecessorId; 
+
+    Node( NodeType nt, NodeId nid_, const RofiWorld& rw, size_t distFromStart_, size_t predecessorId_ ) :
+        nid( nid_ ),
+        world( rw ), 
+        distFromStart( distFromStart_ ),
+        predecessorId( predecessorId_ ) 
+        {
+            Vector cloudEigenVals;
+            switch ( nt ) // each node type stores different information
+            {
+            case NodeType::Shape:
+                shape = rofiWorldToShape( rw );
+                break;
+            case NodeType::Eigen:
+                eigenVals = rofiWorldToEigenValues( rw );
+                break;
+            case NodeType::EigenCloud:
+                shape = rofiWorldToCloud( rw );
+                cloudEigenVals = shape.eigenValues();
+                eigenVals = { static_cast<int>( round( cloudEigenVals(0) * 10000 ) ), 
+                              static_cast<int>( round( cloudEigenVals(1) * 10000 ) ),
+                              static_cast<int>( round( cloudEigenVals(2) * 10000 ) ) };
+                break;
+            default:
+                break;
+            }  
+        }
+};
+
+} // namespace rofi::shapereconfig
+
+namespace rofi::shapereconfig::detail { // auxiliary structs and adjacent configurations generators
+
+using namespace rofi::configuration;
```

Comment:
This will cause problems if something to rofi::configuration is added that collides with your names of functions/classes.